### PR TITLE
Prepend a comment to files setup in container

### PR DESCRIPTION
**Pull request by rhatdan**
setup.rpm attempts to modify /etc/hosts, if it thinks
it has never been modified.  By adding a #comment to the
front of the file, it should prevent this from blowing up
the build.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Attempt to fix https://github.com/containers/buildah/issues/1046

@nalind WDYT?

**Comment by TomSweeneyRedHat** [comment]
@mglantz it would be great if you could try this in your environment to see if it solves your issue.

**Comment by mglantz** [comment]
I can get testing right away @TomSweeneyRedHat, do you have automatic builds including RPM packages or do I need to built it myself?

**Comment by TomSweeneyRedHat** [comment]
@mglantz thanks, no auto builds at this time, you'll need to pull the patch and build/install.

**Comment by mglantz** [comment]
I'm testing, I'll should be done latest tomorrow.

**Comment by mglantz** [comment]
I do not get this to work on a completely clean install of F29 and a build of this fix, same problem as before. I used the same Dockerfile as I used before to test.

**Comment by mglantz** [comment]
I'll try and test on a new EL8 test rig as well later.

**Comment by TomSweeneyRedHat** [comment]
@mairin drats!  But thanks for the test run and I'd be interested in hearing the EL8 test results.

**Comment by TomSweeneyRedHat** [comment]
I'm not at all sure how I got to @mairin on my last reply, but that reply was directed towards @mglantz

**Comment by mglantz** [comment]
I'm waiting for people to sort out availability of -devel packages for EL8 so I can compile the fix. Hopefully I can test today.

**Comment by rhatdan** [comment]
@mglantz Are you saying this fixes the issue or not?

**Comment by rhatdan** [comment]
@mglantz Could you try this one.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat @nalind @vrothberg @giuseppe PTAL

**Comment by mglantz** [comment]
> @mglantz Could you try this one.

I tried it out on a fresh F29 install and I see the same problem I'm afraid. Just in case I'm messing this somehow, full test is as such:
```
cd
cat << 'EOF' >Dockerfile
FROM registry.centos.org/che-stacks/centos-stack-base
RUN sudo yum -y update setup
CMD tail -f /dev/null 
EOF
mkdir buildah
cd buildah
export GOPATH=`pwd`
git clone -b copy1 https://github.com/rhatdan/buildah ./src/github.com/containers/buildah
cd src/github.com/containers/buildah
make
sudo make install
cd
podman build -t buildah:test .
```

**Comment by rhatdan** [comment]
@mglantz Could you run a container and cat /etc/hosts file, to make sure my fix took place.

**Comment by rhatdan** [comment]
Actually you need to do this with buildah bud not with podman.  If we fix it here then we need to vendor this into podman.

File: run.go [hunk 1]
```diff
@@ -638,6 +607,58 @@ func (b *Builder) addNetworkConfig(rdir, hostPath string, chownOpts *idtools.IDP
...
+		if len(values) != 2 {
+			return "", errors.Errorf("unable to parse host entry %q: incorrect format", host)
+		}
+		if values[0] == "" {
+			return "", errors.Errorf("hostname in host entry %q is empty", host)
+		}
+		if values[1] == "" {
+			return "", errors.Errorf("IP address in host entry %q is empty", host)
+		}
+		hosts += fmt.Sprintf("%s\t%s\n", values[1], values[0])
```
Comment (vrothberg) [inline run.go]: Performance nit: string concatenations are very slow. Starting with Go 1.10, there's a `strings.Builder` which is much faster. We could also use a `var buf bytes.Buffer` and `buf.WriteString("...")` which is still magnitudes faster than the `+` operator.

File: run.go [hunk 2]
```diff
@@ -638,6 +607,58 @@ func (b *Builder) addNetworkConfig(rdir, hostPath string, chownOpts *idtools.IDP
...
+	}
+
+	if hostname != "" {
+		hosts += fmt.Sprintf("127.0.0.1   %s\n", hostname)
+		hosts += fmt.Sprintf("::1         %s\n", hostname)
+	}
+	cfile := filepath.Join(rdir, filepath.Base(hostPath))
+	if err = ioutils.AtomicWriteFile(cfile, []byte(hosts), stat.Mode().Perm()); err != nil {
```
Comment (vrothberg) [inline run.go]: I think `bytes.Buffer` is the right type for `hosts`.

File: run.go [hunk 3]
```diff
@@ -638,6 +607,58 @@ func (b *Builder) addNetworkConfig(rdir, hostPath string, chownOpts *idtools.IDP
 	return cfile, nil
 }
 
+// generateHosts creates a containers hosts file
+func (b *Builder) generateHosts(rdir, hostname string, addHosts []string, chownOpts *idtools.IDPair) (string, error) {
+	hostPath := "/etc/hosts"
+	stat, err := os.Stat(hostPath)
```
Comment (vrothberg) [inline run.go]: Can we drop a comment on why the `Stat()` call is necessary? It took me a while to find the user below to preserve the permissions.

**Comment by mglantz** [comment]
> Actually you need to do this with buildah bud not with podman. If we fix it here then we need to vendor this into podman.

@rhatdan Aah. I thought podman called on buildah. Using buildah directly:

```
buildah from registry.centos.org-che-stacks/centos-stack-base
buildah run centos-stack-base-working-container-3 -- sudo yum -y update setup
```
Things works without any problem.

**Comment by rhatdan** [comment]
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1669197

File: run.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply run.go]: Added #1397 to go through and scrub these.  I'm going to let this for now.

File: run.go (hunk 2 above)
Comment (TomSweeneyRedHat) [reply run.go]: @rhatdan thoughts on @vrothberg 's comment here?

Comment (vrothberg) [reply run.go]: @TomSweeneyRedHat, we can defer that to another PR. At some point, I would love to analyze the code for such and similar issues in the hope to further squeeze some cycles :)

Comment (rhatdan) [reply run.go]: Moved to bytes.Buffer,  Is there something better then using the fmt.Sprintf?

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
@vrothberg one last check please and then merge?

**Comment by vrothberg** [comment]
@rh-atomic-bot r+
LGTM

**Pull request closed without merging**